cpp/src/community/detail/mis_impl.cuh

```diff
@@ -0,0 +1,422 @@
+    }
+
+    //
+    // Find maximum rank outgoing neighbor for each vertex
+    // (In case of Leiden decision graph, each vertex has at most one outgoing edge)
+    //
+    auto max_outgoing_rank_id_pairs = allocate_dataframe_buffer<thrust::tuple<weight_t, vertex_t>>(
+      graph_view.local_vertex_partition_range_size(), handle.get_stream());
+
+    per_v_transform_reduce_outgoing_e(
+      handle,
+      graph_view,
+      multi_gpu
+        ? src_rank_cache.view()
+        : detail::edge_major_property_view_t<vertex_t, weight_t const*>(temporary_ranks.data()),
+      multi_gpu ? dst_rank_cache.view()
+                : detail::edge_minor_property_view_t<vertex_t, weight_t const*>(
+                    temporary_ranks.data(), vertex_t{0}),
+      edge_dummy_property_t{}.view(),
+      [] __device__(auto src, auto dst, auto src_rank, auto dst_rank, auto wt) {
+        return thrust::make_tuple(dst_rank, dst);
+      },
+      thrust::make_tuple(std::numeric_limits<weight_t>::lowest(),
+                         invalid_vertex_id<vertex_t>::value),
+      cugraph::reduce_op::maximum<thrust::tuple<weight_t, vertex_t>>{},
+      cugraph::get_dataframe_buffer_begin(max_outgoing_rank_id_pairs));
+
+    // if (debug) {
+    //   CUDA_TRY(cudaDeviceSynchronize());
+    //   std::cout << "Outgoing max neighbor rank-ids: ";
+    //   thrust::for_each(handle.get_thrust_policy(),
+    //                    cugraph::get_dataframe_buffer_cbegin(max_outgoing_rank_id_pairs),
+    //                    cugraph::get_dataframe_buffer_cend(max_outgoing_rank_id_pairs),
+    //                    [] __device__(auto rank_id_tuple) {
+    //                      auto rank = thrust::get<0>(rank_id_tuple);
+    //                      auto id   = thrust::get<1>(rank_id_tuple);
+    //                      printf("\n%d %f\n", id, rank);
+    //                    });
+    // }
+
+    //
+    // Find maximum rank incoming neighbor for each vertex
+    //
+
+    auto max_incoming_rank_id_pairs = allocate_dataframe_buffer<thrust::tuple<weight_t, vertex_t>>(
+      graph_view.local_vertex_partition_range_size(), handle.get_stream());
+
+    per_v_transform_reduce_incoming_e(
+      handle,
+      graph_view,
+      multi_gpu
+        ? src_rank_cache.view()
+        : detail::edge_major_property_view_t<vertex_t, weight_t const*>(temporary_ranks.data()),
+      multi_gpu ? dst_rank_cache.view()
+                : detail::edge_minor_property_view_t<vertex_t, weight_t const*>(
+                    temporary_ranks.data(), vertex_t{0}),
+      edge_dummy_property_t{}.view(),
+      [] __device__(auto src, auto dst, auto src_rank, auto dst_rank, auto wt) {
+        return thrust::make_tuple(src_rank, src);
+      },
+      thrust::make_tuple(std::numeric_limits<weight_t>::lowest(),
+                         invalid_vertex_id<vertex_t>::value),
+      cugraph::reduce_op::maximum<thrust::tuple<weight_t, vertex_t>>{},
```

Comment:
Note that `per_v_tansform_reduce_incoming|outgoing_e` has a bug for min/max reduction. This is fixed in https://github.com/rapidsai/cugraph/pull/3284.